bindings/src/main/java/io/pravega/storage/extendeds3/ExtendedS3ChunkStorage.java

```diff
@@ -0,0 +1,328 @@
+        return true;
+    }
+
+    @Override
+    public boolean supportsTruncation() {
+        return false;
+    }
+
+    //endregion
+
+    //region implementation
+
+    @Override
+    protected ChunkHandle doOpenRead(String chunkName) throws ChunkStorageException, IllegalArgumentException {
+        if (!checkExists(chunkName)) {
+            throw new ChunkNotFoundException(chunkName, "Chunk not found");
+        }
+        return ChunkHandle.readHandle(chunkName);
+    }
+
+    @Override
+    protected ChunkHandle doOpenWrite(String chunkName) throws ChunkStorageException, IllegalArgumentException {
+        if (!checkExists(chunkName)) {
+            throw new ChunkNotFoundException(chunkName, "Chunk not found");
+        }
+        return ChunkHandle.writeHandle(chunkName);
+    }
+
+    @Override
+    protected int doRead(ChunkHandle handle, long fromOffset, int length, byte[] buffer, int bufferOffset) throws ChunkStorageException, NullPointerException, IndexOutOfBoundsException {
+        try {
+            if (fromOffset < 0 || bufferOffset < 0 || length < 0) {
+                throw new ArrayIndexOutOfBoundsException();
+            }
+
+            try (InputStream reader = client.readObjectStream(config.getBucket(),
+                    config.getPrefix() + handle.getChunkName(), Range.fromOffsetLength(fromOffset, length))) {
+                if (reader == null) {
+                    throw new ChunkNotFoundException(handle.getChunkName(), "Chunk not found");
+                }
+
+                int bytesRead = StreamHelpers.readAll(reader, buffer, bufferOffset, length);
+
+                return bytesRead;
+            }
+        } catch (Exception e) {
+            throwException(handle.getChunkName(), "doRead", e);
+        }
+        return 0;
+    }
+
+    @Override
+    protected int doWrite(ChunkHandle handle, long offset, int length, InputStream data) throws ChunkStorageException, IndexOutOfBoundsException {
+        Preconditions.checkArgument(!handle.isReadOnly(), "handle must not be read-only.");
+        try {
+            S3ObjectMetadata result = client.getObjectMetadata(config.getBucket(), config.getPrefix() + handle.getChunkName());
+            client.putObject(this.config.getBucket(), this.config.getPrefix() + handle.getChunkName(),
+                    Range.fromOffsetLength(offset, length), data);
+            return length;
+        } catch (Exception e) {
+            throwException(handle.getChunkName(), "doWrite", e);
+        }
+        return 0;
+    }
+
+    @Override
+    public int doConcat(ConcatArgument[] chunks) throws ChunkStorageException, UnsupportedOperationException {
+        int totalBytesConcated = 0;
+        try {
+            int partNumber = 1;
+
+            SortedSet<MultipartPartETag> partEtags = new TreeSet<>();
+            String targetPath = config.getPrefix() + chunks[0].getName();
+            String uploadId = client.initiateMultipartUpload(config.getBucket(), targetPath);
```

Comment:
Based on our experience with ExtendedS3, there is now a config value `minSizeLimitForConcat`. If size of the source object is below this size then concat is not called. Instead concat is performed using appends.

To avoid confusion, we removed concept of separate "concat using native capabilities" vs "concat using append" from `ChunkStorage`. Instead of that`ChunkedSegmentStorage` itself now decides whether to call concat or not and it directly translates "concat using append" into appropriate write() calls on ChunkStorage .